src/Json/Decode/Applicative.elm

```diff
@@ -0,0 +1,19 @@
+module Json.Decode.Applicative exposing (apply, with)
+
+{-| Json.Decode.Applicative module
+
+This module implements typeclass applicative (apply : m (a -> b) -> a -> b)
+for Decoder functor based on the monadic return (succeed) and bind (andThen)
+
+-}
+
+import Json.Decode exposing (Decoder, andThen, succeed)
+
+
+apply : Decoder (a -> b) -> Decoder a -> Decoder b
+apply df da =
+    andThen (\a -> andThen (\f -> succeed (f a)) df) da
+
+
+with da df =
+    apply df da
```

Comment:
What are your thoughts on `elm-community/json-extra` and https://package.elm-lang.org/packages/elm-community/json-extra/latest/Json-Decode-Extra#andMap ?

Reply:
I think we should consider that the primary goal of this implementation is to show off an understanding of things and thus I would even appreciate "reinventing of the wheel".

Reply:
I certainly could have used that but as @turboMaCk wrote I treated this as an exercise so I wanted to implement the applicative myself (I have actually never implemented Aplicative using Monad so it was a nice exercise for me).

Reply:
we actually do have one implementation of the applicative using monad in our production code even. It's used to implement permutations of values of `List a, List b` to `List (a,b)` :)

Reply:
for curious souls (it's not 100% just an applicative).

```elm
allCacheKeysFor : List DatapointCode -> List AudienceId -> List SegmentId -> List CacheKey
allCacheKeysFor datapointCodes audienceIds segmentIds =
    let
        withMaybe : List a -> (Maybe a -> b) -> List b
        withMaybe xs f =
            if List.isEmpty xs then
                List.singleton <| f Nothing

            else
                List.map (f << Just) xs
    in
    List.map CacheKey datapointCodes
        |> List.andThen (withMaybe audienceIds)
        |> List.andThen (withMaybe segmentIds)
```

so I think seeing you understand similar patterns is valuable for us.